### ISSUE TITLE
NetworkX doesn't automatically import numpy and matplotlib

### ISSUE THREAD

**Issue by mdboom** [iodide-project/pyodide#281]
If a user inports NetworkX and then tries to use its plotting functionality, they get an ImportError about missing matplotlib.

The situation is somewhat complicated.

Numpy and matplotlib are not actually hard dependencies of networkx, since it's possible to do a lot of the graph theoretical things without them. It's only when you try to plot that matplotlib tries to import. It looks like anaconda suffers this same fate if you make a clean environment and put networkx in it. So I'm not sure that putting numpy and matplotlib as dependencies on the package is the right solution.

Unfortunately, the magic that happens to make import matplotlib go out and fetch a package only happens by parsing the initial chunk of code send to pyodide. This is both because (1) being too magical too far down the stack could have a lot of unintended consequences and (2) we can't do an asynchronous import, so we have to do it before running any Python at all.

We could trap the import error and say something helpful like "add import matplotlib to your code to use this feature"...

TL;DR we need some way to make this not so surprising.

See original discussion [here](https://github.com/iodide-project/iodide/pull/1269)

Cc: @hamilton

**Comment by hamilton** [comment]
Hmm, this is indeed tricky!

One possible solution is to separate these two types of dependencies for our purposes. I think it's right that we're not changing the hard dependencies of a library. But If I jump into a notebook expecting to use pyodide, I sort of expect from a workflow point of view that I effectively have a particular set of pre-installed libraries at my disposal. I have to pay a load penalty when I evaluate `import matplotlib` for instance but that's not a big deal - I still know that's part of this distribution, and there is nothing outside of python-land that I need to do to get to work. With that in mind, we could in theory have a separate type of dependency - a soft dependency list, or something like that - which effectively (for now, at least) brings in those packages just like the hard dependency list. But by doing this we're sort of giving the impression of a distribution, and from the user's perspective, it might as well be. After all, that's the common user experience with most anaconda + jupyter builds.

So the upside is the user can do that thing they can do with a lot of other pertinent data science code in pyodide - copy and paste, and it just works. This keeps with our current workflow story with pyodide. The downside is, they'd have to pay the load penalty and it might not even be necessary (though I suppose you can't do much interesting stuff w/o numpy at the very least).

I guess it's a question of what we want the user's mental model to be - do we want user to believe they are starting from a clean environment and do magic + give them tools to build the environment, or do we want them to have the impression that the environment is already available and they can just get to importing?

**Comment by mdboom** [comment]
There's all kinds of complicated solutions we could use to allow the user to choose exactly what they want, but they're all going to end up sort of confusion / user hostile.  I think for now, we can just make `numpy` and `matplotlib` dependencies, and leave a note in the `yaml` that this is different from what upstream does.  We can always revisit if there grows a strong use case to use `NetworkX` without those dependencies.  (I know there are things one can do, but for our environment, I think that's pretty niche).

**Comment by hamilton** [comment]
Yep, mostly concerned with the iodide environment specifically here. Seems like we're going with the latter user mental model, which feels right.
